prometheus/promhttp/http.go

```diff
@@ -406,3 +435,46 @@ func httpError(rsp http.ResponseWriter, err error) {
 		http.StatusInternalServerError,
 	)
 }
+
+// NegotiateEncodingWriter reads the Accept-Encoding header from a request and
+// selects the right compression based on an allow-list of supported
+// compressions. It returns a writer implementing the compression and an the
+// correct value that the caller can set in the response header.
+func NegotiateEncodingWriter(r *http.Request, rw io.Writer, disableCompression bool, compressions []string) (_ io.Writer, encodingHeaderValue string, closeWriter func() error, _ error) {
+	w := rw
+
+	if disableCompression {
+		return w, string(Identity), nil, nil
+	}
+
+	// TODO(mrueg): Replace internal/github.com/gddo once https://github.com/golang/go/issues/19307 is implemented.
+	compression := httputil.NegotiateContentEncoding(r, compressions)
+
+	switch compression {
+	case "zstd":
+		// TODO(mrueg): Replace klauspost/compress with stdlib implementation once https://github.com/golang/go/issues/62513 is implemented.
+		z, err := zstd.NewWriter(rw, zstd.WithEncoderLevel(zstd.SpeedFastest))
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		z.Reset(w)
+		w = z
+
+		return w, compression, z.Close, nil
+	case "gzip":
+		gz := gzipPool.Get().(*gzip.Writer)
+		defer gzipPool.Put(gz)
```

Comment:
This has to be done on Close. It's a subtle bug, but concurrent Negotiate calls might corrupt gzip writers (hard to repro on tests I guess). "Put" means "I won't use gz anymore", which will happen on return of this function. But that's not true is it?

Reply:
Good catch! I was wondering about that as well. Thanks for the suggestions on returning a function here that calls both. TIL :)

Reply:
I think you accidently left this defer here though.